src/block.c

```diff
@@ -242,81 +242,76 @@ struct block_device *block_device_init(const struct list *cntrl_list, const char
 	char link[PATH_MAX];
 	char *host = NULL;
 	struct block_device *device = NULL;
-	struct pci_slot *pci_slot = NULL;
-	send_message_t send_fn = NULL;
-	flush_message_t flush_fn = NULL;
 	int host_id = -1;
-	char *host_name;
+	char *host_name = NULL;
 
-	if (realpath(path, link)) {
-		cntrl = block_get_controller(cntrl_list, link);
-		if (cntrl != NULL) {
-			pci_slot = vmdssd_find_pci_slot(cntrl->ctx, link);
-			if (cntrl->cntrl_type == LED_CNTRL_TYPE_VMD && !pci_slot)
-				return NULL;
-			host = _get_host(link, cntrl);
-			if (host == NULL)
-				return NULL;
-			host_name = get_path_hostN(link);
-			if (host_name) {
-				if (sscanf(host_name, "host%d", &host_id) != 1)
-					host_id = -1;
-				free(host_name);
-			}
-			flush_fn = _get_flush_fn(cntrl, link);
-			send_fn = _get_send_fn(cntrl, link);
-			if (send_fn  == NULL) {
-				free(host);
-				return NULL;
-			}
-		} else {
-			return NULL;
-		}
+	if (!realpath(path, link))
+		goto error;
 
-		device = calloc(1, sizeof(*device));
-		if (device) {
-			struct _host_type *hosts = cntrl ? cntrl->hosts : NULL;
-			device->cntrl = cntrl;
-			device->sysfs_path = strdup(link);
-			if (!device->sysfs_path) {
-				free(host);
-				free(device);
-				return NULL;
-			}
-			device->cntrl_path = host;
-			device->ibpi = IBPI_PATTERN_UNKNOWN;
-			device->ibpi_prev = IBPI_PATTERN_NONE;
-			device->send_fn = send_fn;
-			device->flush_fn = flush_fn;
-			device->timestamp = timestamp;
-			device->host = NULL;
-			device->host_id = host_id;
-			device->encl_index = -1;
-			device->raid_dev = NULL;
-			while (hosts) {
-				if (hosts->host_id == host_id) {
-					device->host = hosts;
-					break;
-				}
-				hosts = hosts->next;
-			}
-			if (cntrl && cntrl->cntrl_type == LED_CNTRL_TYPE_SCSI) {
-				device->phy_index = cntrl_init_smp(link, cntrl);
-				if (!dev_directly_attached(link)
-						&& !scsi_get_enclosure(cntrl->ctx, device)) {
-					lib_log(cntrl->ctx, LED_LOG_LEVEL_DEBUG,
-						"Device initialization failed for '%s'", path);
-					free(device->sysfs_path);
-					free(device->cntrl_path);
-					free(device);
-					device = NULL;
-				}
-			}
-		} else if (host) {
-			free(host);
+	cntrl = block_get_controller(cntrl_list, link);
+	if (!cntrl)
+		goto error;
+
+	if (cntrl->cntrl_type == LED_CNTRL_TYPE_VMD &&
+			!vmdssd_find_pci_slot(cntrl->ctx, link))
+		goto error;
+	host = _get_host(link, cntrl);
+	if (host == NULL)
+		goto error;
+	host_name = get_path_hostN(link);
+	if (host_name) {
+		if (sscanf(host_name, "host%d", &host_id) != 1)
+			host_id = -1;
+		free(host_name);
+	}
+
+	device = calloc(1, sizeof(*device));
+	if (!device)
+		goto error;
+
+	struct _host_type *hosts = cntrl ? cntrl->hosts : NULL;
```

Comment:
Done

Reply:
> I proposed to make it configurable via parameter passed to scan:

```c
led_status_t LED_SYM_PUBLIC led_scan(struct led_ctx *ctx, bool scan_slots);
```

> For reset and next we will return error if  ctx->sl.slot_list.head is NULL.
My goal is to left all scans in led_scan() because I want highlight that it is the one place when we are iterating over the system (sysfs, ioctls and others).
The later we should impose led states only, assuming that we have all necessary information to do it.
Let we know how you see it.

I actually prefer what I did the first time.  Where the API caller called into the API and got an abstraction of a list and then iterated on it.  This was done as we don't want to expose a linked list implementation.

Regardless, there is no sysfs `scan` performed in either the implementations I've provided.  It simply walks over the in memory representation of what the scan already did and presents it with opaque data types.

Before I re-write this for the third time, please consider what the API end user experiences, how it can be used, the limitations of it etc.  For example, in the first implementation a user could scan, get the list, close the library handle and continue to use the list abstraction indefinitely.  The API user could also have more than one at a time too, representing change over time.  Thus for example, the functionality where ledmon keeps track of state and, adds/updates/removes could be assisted by what the API provides.

Reply:
> Before I re-write this for the third time, please consider what the API end user experiences, how it can be used, the limitations of it etc. For example, in the first implementation a user could scan, get the list, close the library handle and continue to use the list abstraction indefinitely. The API user could also have more than one at a time too, representing change over time. Thus for example, the functionality where ledmon keeps track of state and, adds/updates/removes could be assisted by what the API provides.

I believe that context is always must have. The main reason is:
- Always use a library context object. Every thread should then
    operate on its own context. (https://git.kernel.org/pub/scm/linux/kernel/git/kay/libabc.git/plain/README)

That is the guidance I would like to follow to limit complexity for now but maybe I've taken in too serious. I understand that the approach you proposed is more flexible and could save some memory. To achieve that, we can handle two contexts and compare them. Nowadays we don't need to save every possible memory cell :)
The main purpose of the library is to allow led patterns manipulation. To achieve that context is always needed.

I see your first approach could be valuable for ledmon where we need to compare two system scans. I think that we can handle that in separate pull If we will determine that there is real use case for that because the first question we should ask ourselves is: Do we have user for the new API?

First, I would like to focus on removing ledmon code from lib and we agreed to handle that later. Then, we will know which list can be eventually coped for the user and if it is really needed (we have request or need to do it).

In this comment I wanted to discuss moving slots list generation to led_scan(), the discuss goes into other direction.
Let me know how you see this problem.

Thanks for giving you personal PoV- I really appreciate that.

Reply:
> > Before I re-write this for the third time, please consider what the API end user experiences, how it can be used, the limitations of it etc. For example, in the first implementation a user could scan, get the list, close the library handle and continue to use the list abstraction indefinitely. The API user could also have more than one at a time too, representing change over time. Thus for example, the functionality where ledmon keeps track of state and, adds/updates/removes could be assisted by what the API provides.
> 
> I believe that context is always must have. The main reason is:
> 
>     * Always use a library context object. Every thread should then
>       operate on its own context. (https://git.kernel.org/pub/scm/linux/kernel/git/kay/libabc.git/plain/README)
> 
> 
> That is the guidance I would like to follow to limit complexity for now but maybe I've taken in too serious. I understand that the approach you proposed is more flexible and could save some memory. To achieve that, we can handle two contexts and compare them. Nowadays we don't need to save every possible memory cell :) The main purpose of the library is to allow led patterns manipulation. To achieve that context is always needed.

It's not two contexts, it's only one context and 1 or more list abstractions.  If you look at the libabc implementation:
https://git.kernel.org/pub/scm/linux/kernel/git/kay/libabc.git/tree/src/abc/libabc.h  you'll see that I followed it in the first implementation.  You'll also see that I didn't implement the reference counting semantics.  I've avoided that as I've had some bad experiences working with memory leaks with reference counting interfaces, but we could always add that at some point too.

> I see your first approach could be valuable for ledmon where we need to compare two system scans. I think that we can handle that in separate pull If we will determine that there is real use case for that because the first question we should ask ourselves is: Do we have user for the new API?

As soon as we have a released and packaged version of the led library I'm going to utilize it in libstoragemgmt, to replace the SES code it already has.  I'm also going to extend the libstoragemgmt API by exposing the slot identification.  This is for ceph consumption.  I have an open RFE for adding functionality like this ref. https://bugzilla.redhat.com/show_bug.cgi?id=1921303 .  I also have a local PR which changes the slot identifier for SES to one that will be accurate over reboots & dynamic reconfiguration by utilizing the enclosure SN for part of the slot ID.  I've delayed introducing that as it's an enhancement that can be done at anytime.

> First, I would like to focus on removing ledmon code from lib and we agreed to handle that later. Then, we will know which list can be eventually coped for the user and if it is really needed (we have request or need to do it).
> 
> In this comment I wanted to discuss moving slots list generation to led_scan(), the discuss goes into other direction. Let me know how you see this problem.

Like I mentioned before.  IMHO, the scan walks the system, the iterators for systems & slots simply wrap what we have in an abstraction for API user consumption.  I would prefer to go back to my initial abstraction as I believe it was more flexible and from an API users perspective more intuitive.  It also only consumes additional memory if a library user chooses to use that part of the API.

> Thanks for giving you personal PoV- I really appreciate that.

Thank you for being open to a discussion.  The public API and how it works really is the important part.  The implementation can change, but the public API will remain.

Reply:
> It's not two contexts, it's only one context and 1 or more list abstractions. If you look at the libabc implementation:

I meant that now (with approach after rework, let say v2) you can do this using 2 contexts.

> https://git.kernel.org/pub/scm/linux/kernel/git/kay/libabc.git/tree/src/abc/libabc.h you'll see that I followed it in the first implementation. You'll also see that I didn't implement the reference counting semantics. I've avoided that as I've had some bad experiences working with memory leaks with reference counting interfaces, but we could always add that at some point too.

I tried to determine how the list interface is implemented and I failed to find proper functions in [libabc.c](https://git.kernel.org/pub/scm/linux/kernel/git/kay/libabc.git/tree/src/libabc.c).
How the list is retuned? Is it pointer or copy of original list? 

I agree, we don't need reference counting now.

> Like I mentioned before. IMHO, the scan walks the system, the iterators for systems & slots simply wrap what we have in an abstraction for API user consumption. I would prefer to go back to my initial abstraction as I believe it was more flexible and from an API users perspective more intuitive. It also only consumes additional memory if a library user chooses to use that part of the API.

OK, we have two problems here:
- cloning the list (if I remember correctly that was how you implemented it first time, correct it if I'm wrong)
- making iterator public

We should avoid cloning the list, it seems to not be natural and should not be part of library. Every caller can make the list copy on its own way. Do you agree here?

For second problem- we should make iterator accessible for caller. Approach prosed by me is not good. I can see two cases where access to iterator in necessary:
- nested list iteration (for in for)
We can make that really simple if we have access to iterators
- get_prev method
Not needed now but It could be needed in the future.

And as you said- it is more intuitive. I have strong feeling that I don't remember correctly your first implementation and that is all based on my incorrect assumption that you returned copy of list to the caller. I cannot verify that after squashing. I'm sure that I saw it somewhere but I don't remember exactly where.

Reply:
> The code was walking the systems/slots and allocates an internal structure for each item in the list. It only contains the information presented in the public API, so it's much smaller than what is present internally. Its not a clone of anything. One of the main reasons this is done is that would be very difficult to handle the case where a user gets the iterator, and during iteration they do a `led_scan`. That could potentially invalidate one or more internal structures if we simply provide iterating over the internal data structures if devices are added/removed.

You right, it is not exact clone, it is a new list made from original one. "Clone" was not perfect wording for that. 

IMHO we should accept the risk you mentioned. I think that it is not a problem. User should know that after led_scan() pointers will be invalidated and if user need old list then he must make a list copy itself. The solution for the problem should stay in application, not in library.
The user should decide whose properties are needed, and should be copied to the internal temporary list. All we need is to provide API to access them. It also gives a direct control of allocated memory (application allocates, application frees).

> I'm not sure what the problem is with making an iterator public. This is simply a place holder for their existing location in the internal list. The main point of exposing iteration is we then don't expose any kind of linked list structure(s). Iterators can be used across different languages bindings too.

Sorry wrong word. I meant "something what need to be resolved/ addressed", maybe concepts / ideas were better.  Sorry for confusing you.

> I'm confused by this. The public API as access to iterator, eg. `next`. We can easily add a `previous` and `reset` to what I initially did. If the user wanted to do nested for loops they would need to get two different list instances. At the moment I'm struggling to see the use case, but they could do it if they needed to.
> 
> > And as you said- it is more intuitive. I have strong feeling that I don't remember correctly your first implementation and that is all based on my incorrect assumption that you returned copy of list to the caller. I cannot verify that after squashing. I'm sure that I saw it somewhere but I don't remember exactly where.
> 
> I still have an old branch. I can resurrect and implement `previous` and `reset`. I'll also include a client example showing its use. I have one already that I'm using for testing.

You don't need to implement functionalities that are not needed. The goal of my comment was to find examples where public iterator brings are real advantage, even if they are not needed now.

Conclusion: I agree that your first approach were better but I would like to not use "internal structure for each item" just present everything as is and document that pointers will be invalidated after led_scan().

Reply:
> Conclusion: I agree that your first approach were better but I would like to not use "internal structure for each item" just present everything as is and document that pointers will be invalidated after led_scan().

I've modified the PR.  The controller list abstraction is using the internal data structures without making a copy.  This was pretty straight forward and the code is actually smaller.  However, the slot data is not so easy.

1. Some of the information is generated, it doesn't exist in any current data structures like the slot id and the current LED status which needs to be generated/retrieved.  Thus if we don't take the current approach of allocating a list of structures which store this information when we iterate through the code to retrieve it, we are then forced to require the API user to check a status code during list iteration to see if any errors occurred as the data would need to be retrieved during the `next` or `previous` iteration.  From an API user perspective, it's much nicer to retrieve the list and check the status once that all is good and then simply walk the data without having to constantly check if it's ok.
2. In my first implementation the API user would specify the controller to retrieve slots for.  Later I changed this to simply retrieve all slots on the system as the returned slot information includes the enumerated controller type.  From an API user perspective I believe this is better.  Otherwise the API user needs to loop across the different supported controllers that support the slots API and retrieve them individually.  To do this iteratively without allocating a new list structure will require some really interesting code as we will need to internally iterate over three different linked lists, each of which contains dissimilar data.

I'll continue to investigate ideas on how to implement this without it being a significant increase in complexity, but I'm not seeing it at the moment.

Reply:
Do not fear to make a slots list inside the context to contain all necessary properties -that fits the design I want.
I don't want to return copied list, with no reference inside the context. If you need to create additional list to combine all necessary properties for slots- that is fine.

Reply:
> Do not fear to make a slots list inside the context to contain all necessary properties -that fits the design I want. I don't want to return copied list, with no reference inside the context. If you need to create additional list inside context to combine all necessary properties for slots- that is fine.

I'm not exactly sure what you're saying here.  Regardless, I'm going to leave this PR the way it is as I don't think I can reduce the list elements for listing slots.  I've completed a different implementation that generates the slot data during iteration and placed it in this branch: https://github.com/tasleson/ledmon/tree/library_rfc_less_alloc . This branch isn't complete, as I would want to change the signature of the controller iteration to report back status during each iteration too.

Please take a look at both and let me know which direction you prefer.

Reply:
> I'm not exactly sure what you're saying here.

Apologize for being inaccurate. Probably the root cause it that I'm only the reviewer.  I will analyze what you prepared and eventually I will come with my code, to not bother you more with more comments. Then we will choose the way.

> Please take a look at both and let me know which direction you prefer.

I need few days to do that.